lib/arborist/index.js

```diff
@@ -122,28 +126,40 @@ class Arborist extends Base {
       }
     }
     for (const extra of extraneous) {
-      set.add(extra)
+      wsDepSet.add(extra)
     }
 
-    return set
+    return wsDepSet
   }
 
-  excludeWorkspacesDependencySet (tree) {
-    const set = new Set()
-    for (const edge of tree.edgesOut.values()) {
-      if (edge.type !== 'workspace' && edge.to) {
-        set.add(edge.to)
-      }
-    }
-    for (const node of set) {
-      for (const edge of node.edgesOut.values()) {
-        if (edge.to) {
-          set.add(edge.to)
-        }
-      }
-    }
+  rootDependencySet (tree) {
+    const rootDepSet = new Set()
+    depth({
+      tree,
+      visit: node => {
+        node.edgesOut.forEach(({ to }) => {
+          if (!to) {
+            return
+          }
+          to.edgesIn.forEach(edgeIn => {
+            if (!to.isWorkspace && (
+              edgeIn.from.isRoot || rootDepSet.has(edgeIn.from))
+            ) {
+              rootDepSet.add(to)
+            }
+          })
+        })
+        return node
+      },
+      filter: node => node,
+      getChildren: (node, tree) =>
+        [...tree.edgesOut.values()].map(edge => edge.to),
+    })
+    return rootDepSet
+  }
 
-    return set
+  excludeWorkspacesDependencySet (tree) {
```

Comment:
Why this extra function?

Reply:
Added as an alias, as I think the original name of the function could be misleading. It only makes sense if you're talking about it as a filter set, which is the intention, so I added an alias to what the set actually contains. It's very confusing. Maybe some comments would be enough.